### PR TITLE
feat: quality-of-life UI improvements and misc changes

### PR DESCRIPTION
**Pull request by alexispurslane**
## Please make sure to check the following tasks before opening and submitting a PR

* [x] I understand and have followed the [contribution guide](https://github.com/revoltchat/revolt/discussions/282)
* [x] I have tested my changes locally and they are working as intended
* [x] These changes do not have any notable side effects on other Revolt projects

## Features

- Scroll to bottom on new messages (or when keyboard opens)
- Render messages from bottom to top like Discord
- Render channel header at the start of any channel or direct message (or saved notes), again ala Discord
- Add a UI setting for font size that changes the message font size and the message box font size (as well as adjusting the padding in various places to make this look nice)
- Reintroduce the AMOLED theme with some extra UI visual hints to make it more usable
- Remember the last server or direct message the user was in and restore them at startup. 
- Remember, for each server, the last channel in that server the user was in, and restore that channel when visiting that server (including on startup)

## Misc Changes
- Remove some dead code (including uncommented code that was nevertheless never used)
- Make Gradle build a full app and an F-Droid-compatible set of binaries.

### CODE REVIEW

**Comment by alexispurslane** [comment]
For my next PR I'm going to try to make the process of restoring your chat location faster, by loading it with the UI instead of loading the UI and *then* changing the state *again*, I'm going to try to fix the message grouping by time, and I'm going to try to make messages just stay in frame without needing to auto scroll every time

**Comment by alexispurslane** [comment]
Question: many of the components in this application currently are far too gigantic and mixed together in functionality to be manageable, it makes it very difficult to find where anything is done or make changes. Would you be open to accepting a pool request from me refactoring a lot of the gigantic 500 line components into subcomponents? Most of the sub components probably won't be reused in any way but I still think breaking it up and separating the concerns a little bit will make it more readable and easy to change.

**Comment by Rexogamer** [comment]
I have been moving to refactor/split things more - I would prefer this be done somewhat gradually, but I would be open to PRs which help with this effort.

Now would also be a good time to mention that I've been moving to a new message view using a flat list instead (currently accessible via the Debug page on the DM list) - as such, while these changes are fine (incoming review comments aside), I wouldn't necessarily spend *too* much time on the current one at this point.

**Comment by alexispurslane** [comment]
ah well, looks like I'm not working on it today. Gradle is shitting itself even though I've changed nothing about my build environment so I can't get it to even build the app and i really don't feel like dealing with its bullshit (I spent six hours trying to get it to work the first time)

**Comment by Rexogamer** [comment]
No rush - I'll test these changes locally and post some review comments soon, but you can take your time ^^

**Review by Rexogamer — CHANGES_REQUESTED** [review]
- Last channels feature works without any apparent issues
- Setting remarks LGTM/will be useful
- Mentions/size changing look fine minus the review comments

File: src/Generic.tsx [hunk 1]
```diff
@@ -182,7 +184,14 @@ export const app = {
         key: 'ui.messaging.messageSpacing',
         category: 'appearance',
         name: 'Message spacing',
-        default: '3',
+        default: '15',
```
[the remark this answers is not in the document]
Comment (alexispurslane) [reply src/Generic.tsx:187]: Sorry, yeah 😅 I struggle with visual processing due to an injury and set all my font sizes huge. The one I actually use in the app is 18!

File: src/MessageView.tsx [hunk 2]
```diff
@@ -340,19 +338,33 @@ export class Messages extends React.Component {
...
+            }}>
+            <Text style={{...styles.loadingHeader, textAlign: 'left'}}>
+              {this.props.channel?.channel_type === 'DirectMessage'
+                ? '@' + this.props.channel?.recipient?.username
+                : this.props.channel?.channel_type === 'SavedMessages'
+                ? 'Saved Notes'
+                : '#' + this.props.channel?.name}
+            </Text>
+            <Text style={styles.beginningRemark}>
+              This is the start of your conversation.
```
[the remark this answers is not in the document]
Comment (alexispurslane) [reply src/MessageView.tsx:361]: Hmm. Having it is really nice for me for purely psychological reasons but it does make sense to comment it out for now. We really need a reliable way of figuring out if we are at the first and last messages — speaking of which, why don't the messages that come through from the request that we use to check if we're at the first message have IDs???

**Comment by alexispurslane** [comment]
Okay in a few days when I'm feeling better I'll get on these changes. I also have some improvements I want to try to rhe last channel code

File: src/MessageView.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Rexogamer) [reply src/MessageView.tsx:361]: That sounds odd - I haven't seen any instances of messages without IDs? `ReplyingMessage`s have the actual message object as a property, but those shouldn't be relevant here so 🤔

[the remark this answers is not in the document]
Comment (alexispurslane) [reply src/MessageView.tsx:361]: I can't remember super clearly rn, but in the place where we request new messages for a channel, either the messages that are requested or the final message that's stored in the asynchronous storage object doesn't have an ID, which means it's not possible to tell whether the user has scrolled down to the very last message in the channel. Which is why I think you had setting the final message flag commented out

[the remark this answers is not in the document]
Comment (Rexogamer) [reply src/MessageView.tsx:361]: I think that was written/commented out by Tai? Either way, when you can get the build to work I'd be grateful for any info you can find here

**Comment by alexispurslane** [comment]
@Rexogamer I figured out what was wrong with atLatestMessages, the base message object was just being wrapped in order to add grouping information so I had to do `.message._id`. So I'm adding in the atLatestMessages logic to make scrolling down work. But I've noticed that since the new commits, `bottomOfPage` no longer works, which means a couple other scrolling related things (including my logic to make sure that if the user has manually scrolled up, that when a new message is added they don't get scrolled down, which is something you noticed) don't work now, so now I've got to track down that issue.

**Comment by Rexogamer** [comment]
I'm tempted to suggest splitting this PR up a bit - most of this (the last channel/server feature, the font size setting, the setting remarks, the Gradle tweaks, the design updates and the AMOLED re-enablement) could all be merged now, while the stuff that seems to be needing more investigation (I'm assuming just the scrolling changes?) could be split off into its own PR to not hold the rest of it up.

**Comment by alexispurslane** [comment]
> I'm tempted to suggest splitting this PR up a bit - most of this (the last channel/server feature, the font size setting, the setting remarks, the Gradle tweaks, the design updates and the AMOLED re-enablement) could all be merged now, while the stuff that seems to be needing more investigation (I'm assuming just the scrolling changes?) could be split off into its own PR to not hold the rest of it up.

That's a good idea, it would just be a lot of work to do so, and I'm starting to despair of this entire project because I've spent about four hours today trying to get/keep the react native build system even working at all, and the app (with no changes on my end) has decided to completely stop running at all just recently, and nothing I can do seems to change it. Plus, I pulled your upstream changes, and there's several unhandled promise errors every few seconds and I can't get a stack trace on them that even ends anywhere in the code at all. I just really hate react native tbh. I'm usually far more productive than this but there's no room for me to be productive when I have to spend all day dealing with flimsy webdev bullshit (sorry for the cursing)

**Comment by alexispurslane** [comment]
I hate this ******* piece of **** so much react native can suck my ****

**Comment by Rexogamer** [comment]
It's fine - I can create a new branch and revert any changes I would rather not merge. I'm slightly confused, however - I haven't seen any unusual errors on my end? React Native can be a pain, certainly

**Pull request closed without merging**